Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -523,6 +523,93 @@ A similar approach works. Answer the following questions:
+-rw-r--r--  1 _installd _installd 189K Nov 16 06:37 libswiftDispatch.dylib
+-rw-r--r--  1 _installd _installd 1.9M Nov 16 06:38 libswiftFoundation.dylib
+-rw-r--r--  1 _installd _installd  76K Nov 16 06:37 libswiftObjectiveC.dylib
+```
+
+Looking at these _traces_ that Frida _lefts behind_, you might already imagine that detecting those would be a trivial task. And actually, so trivial will be bypassing that detection. But things can get much more complicated. The following table shortly presents a set of some typical Frida detection methods and a short discussion on their effectiveness.
+
+> Some of the following detection methods are implemented in the [iOS Security Suite](https://github.com/securing/IOSSecuritySuite "iOS Security Suite").
+
+| Method | Description | Discussion |
+| --- | --- | --- |
+| **Check Signature of an App** | One detection mechanism would be to verify the signature of the app. In order to embed the frida-gadget within the IPA, it would need to be repackaged and resigned. There is a library available in Objective-C called [iOS-App-Security-Class](https://github.com/karek314/iOS-App-Security-Class "iOS-App-Security-Class") that checks if the app is encrypted or not, by verifying `cryptid`. If this value is 0, the app was dumped (e.g. "clutched") and the Fairplay DRM encryption is not present anymore.  | This indicates that someone is trying to tamper and reverse engineer an app. There are limitations in such an implementation, as you cannot easily test if the check for `cryptid` is properly working. The app will only be encrypted by Apple after you submitted it to the App Store. |
+| **Check The Environment For Related Artifacts**  | Artifacts can be packaged files, binaries, libraries, processes, and temporary files. For Frida, this could be the frida-server running in the target (jailbroken) system (the daemon responsible for exposing Frida over TCP) or the frida libraries loaded by the app. | Inspecting running services is not possible for an iOS app on a non-jailbroken device. The Swift method [CommandLine](https://developer.apple.com/documentation/swift/commandline "CommandLine") is not available on iOS to query for information about running processes, but there are inofficial ways, like using [NSTask](https://stackoverflow.com/a/56619466 "How can I run Command Line commands or tasks with Swift in iOS?"). Nevertheless when using this method, the app will be rejected during the App Store review process. There is no other public API available to query for running processes or execute system commands within an iOS App. Even if it would be possible, bypassing this would be as easy as just renaming the corresponding Frida artifact (frida-server/frida-gadget/frida-agent). Another way to detect Frida, would be to walk through the list of loaded libraries and check for suspicious ones (e.g. those including "frida" in their names), which can be done by using `_dyld_get_image_name`.
```

Comment:
```suggestion
| **Check The Environment For Related Artifacts**  | Artifacts can be packaged files, binaries, libraries, processes, and temporary files. For Frida, this could be the frida-server (the daemon responsible for exposing Frida over TCP) running in the target (jailbroken) system or the frida libraries loaded by the app. | Inspecting running services is not possible for an iOS app on a non-jailbroken device. The Swift method [CommandLine](https://developer.apple.com/documentation/swift/commandline "CommandLine") is not available on iOS to query for information about running processes, but there are unofficial ways, such as using [NSTask](https://stackoverflow.com/a/56619466 "How can I run Command Line commands or tasks with Swift in iOS?"). However, when using this method, the app will be rejected during the App Store review process. There is no other public API available to query for running processes or execute system commands within an iOS App. Even if it would be possible, bypassing this would be as easy as just renaming the corresponding Frida artifact (frida-server/frida-gadget/frida-agent). Another way to detect Frida, would be to walk through the list of loaded libraries and check for suspicious ones (e.g. those including "frida" in their names), which can be done by using `_dyld_get_image_name`.
```